sound/soc/sof/ipc4-pcm.c

```diff
@@ -13,6 +13,46 @@
 #include "ipc4-priv.h"
 #include "ipc4-topology.h"
 
+static int sof_ipc4_set_multi_pipeline_state(struct snd_sof_dev *sdev, u32 state,
+					     struct snd_sof_pipeline_trigger_info *info)
+{
+	struct ipc4_pipeline_set_state_data data;
+	struct sof_ipc4_msg msg = {{ 0 }};
+	u32 primary, ipc_size;
+	int i;
+
+	primary = state;
+	primary |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_GLB_SET_PIPELINE_STATE);
+	primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
+	msg.primary = primary;
+
+	msg.extension = SOF_IPC4_GLB_PIPE_MULTI;
+
+	data.count = 0;
+	/*
+	 * IPC4 requires pipelines to be triggered in order starting at the sink and
+	 * walking all the way to the source. So add the pipeline instance ID's starting with the
+	 * last one in the trigger list.
+	 */
+	for (i = info->count - 1; i >= 0; i--) {
+		struct snd_sof_widget *pipe_widget = info->pipelines[i];
+		struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
+		/* skip triggering backend widgets */
+		if (!pipeline->is_backend) {
+			data.pipeline_ids[i] = info->pipelines[i]->instance_id;
+			data.count++;
+		}
```

Comment:
@ranj063, the issue is that you record _all_ pipelines, including the BE one to the `pipeline_info` and when you have two pipelines in the route (one FW and one BE, capture) you would end up doing:
```
[ 8443.720222] sof_ipc4_set_multi_pipeline_state: info->count: 2
[ 8443.720226] sof_ipc4_set_multi_pipeline_state: i: 1, data.count: 1 (is_backend: 0)
[ 8443.720230] sof_ipc4_set_multi_pipeline_state: i: 0, data.count: 1 (is_backend: 1)
[ 8443.720234] sof_ipc4_set_multi_pipeline_state: data.count: 1
```
Put the FE pipeline at `data.pipeline_ids[1]` and leave the `data.pipeline_ids[0]` untouched, but you still tell the firmware that there is one pipeline to multi-set-state...

In `sof_set_up_widgets_in_path()` the comment say that the pipelines are added to the list in source to sink order, right?
Here you want to revert that order, if I read the comment correctly.
This diff fixes the issue, but I'm not sure if this is what you want:
```diff
diff --git a/sound/soc/sof/ipc4-pcm.c b/sound/soc/sof/ipc4-pcm.c
index 85a96b75ebb7..b20c61a02d2b 100644
--- a/sound/soc/sof/ipc4-pcm.c
+++ b/sound/soc/sof/ipc4-pcm.c
@@ -40,10 +40,8 @@ static int sof_ipc4_set_multi_pipeline_state(struct snd_sof_dev *sdev, u32 state
 		struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
 
 		/* skip triggering backend widgets */
-		if (!pipeline->is_backend) {
-			data.pipeline_ids[i] = info->pipelines[i]->instance_id;
-			data.count++;
-		}
+		if (!pipeline->is_backend)
+			data.pipeline_ids[data.count++] = info->pipelines[i]->instance_id;
 	}
 
 	ipc_size = sizeof(u32) * (data.count + 1);

```

Reply:
@ujfalusi I messed this one royally when I switched the order of iterating over the array. fixed now. Thanks!